DRILL-5721: Query with only root fragment and no non-root fragment hangs when Drillbit to Drillbit Control Connection has network issues

sohami opened the pull request.

sohami commented:
@parthchandra - Please help to review this PR.

sohami commented:
Rebased the branch to latest master and updated test's included in DRILL-5701

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java:
```diff
@@ -1073,26 +1070,22 @@ public QueryId getQueryId() {
+    queryManager.addFragmentStatusTracker(rootFragment, true);
+
+    // FragmentManager is setting buffer for FragmentContext
+    if (rootContext.isBuffersDone()) {
```
Inside AbstractFragmentManager constructor, since that's where the Incoming Buffers are created.

https://github.com/apache/drill/pull/919/files#diff-d7418a1fd07314f268049ae9f07b7ed3

```
+  public AbstractFragmentManager(final PlanFragment fragment, final FragmentExecutor executor, final FragmentStatusReporter statusReporter, final FragmentRoot rootOperator) {
 +    this.fragmentHandle = fragment.getHandle();
 +    this.fragmentContext = executor.getContext();
 +    this.buffers = new IncomingBuffers(fragment, fragmentContext);
 +    this.fragmentContext.setBuffers(buffers);
```

parthchandra replied:
Yes but from the looks of it the fragment manager hasn't been created yet. In fact it seems it is created in the else part of the condition you're checking.

sohami replied:
Sorry about the confusion. It was moved out of else in second commit.

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java:
```diff
@@ -1109,7 +1102,7 @@ private void setupNonRootFragments(final Collection<PlanFragment> fragments) thr
       return;
     }
-    /*
+    /*z
```
It was part for refactor commit only and was later removed in next commit.

parthchandra commented:
+1 LGTM

sohami commented:
@parthchandra - thanks for the review!

paul-rogers commented:
@sohami, can you clean up the stray commits and ask @parthchandra to again review the later commits?

sohami commented:
@paul-rogers - Not sure which commit you are referring to. All the commits are reviewed by @parthchandra. I just rebased commits to latest master after last week pull request since @amansinha100 was observing some random functional test failure. After rebase I ran those tests in isolation for multiple iterations and those were passing. Let me know if you want me to combine all these 3 commits.

paul-rogers commented:
OK, thanks. Got confused by the commit messages that appeared after the +1. Maybe in the future, when we rebase, just add a comment saying that's what happened.

paul-rogers commented:
Tried to commit this one to master. The functional test run failed with 20+ "random failures." That is too many to comfortably allow. So, @sohami, please rerun this, determine if the 20+ failures are, indeed, random or due to this change, and post a note here when the PR is ready for another attempt at commit into master.

sohami commented:
@paul-rogers  - Last week @amansinha100 also reported 5 Random failures happening with this commit. But when ran in isolation for multiple iterations those tests were passing all the time. Can you please share the list of tests failing for you with this commit ? I can try to run a sample of those tests in isolation and see if those are really random.

sohami commented:
I rebased this PR to latest master and [re-ran functional test](http://10.10.104.91:8080/job/Functional/2447/console) and seeing only 2 random failures now. As per @agirish recommendation I have created JIRA's to track these random issues. [DRILL-5788](https://issues.apache.org/jira/browse/DRILL-5788) & [DRILL-5789](https://issues.apache.org/jira/browse/DRILL-5789).

paul-rogers commented:
Attempt to merge into master failed:

```
CONFLICT (content): Merge conflict in exec/java-exec/src/main/java/org/apache/drill/exec/work/fragment/FragmentStatusReporter.java
```

Please rebase onto latest master. While at it, please also squash commits.

sohami commented:
Rebased on latest master and squashed the initial 3 commits. But I have kept the commit to resolve conflict separate as there are some changes made w.r.t DRILL-3449 behavior, and added some new unit tests.
